### PR TITLE
Django SMS Toolkit V1

### PR DESCRIPTION
**Pull request by alicertel**

### CODE REVIEW

File: django_sms_toolkit/tasks.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+from twilio.base.exceptions import TwilioRestException
+from twilio.rest import Client
+
+from .models import TwilioMessage
+from .settings import DJANGO_SMS_TOOLKIT_SETTINGS
+
+
+@shared_task(retry_backoff=30, retry_kwargs={'max_retries': 3})
+def send_sms(from_number, to_number, body, recipient_id=None):
+    if not DJANGO_SMS_TOOLKIT_SETTINGS["SEND_SMS"]:
```
Comment (efe) [inline django_sms_toolkit/tasks.py]: I think we should never send SMS while testing.

File: django_sms_toolkit/tasks.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    # Twilio does not accept body larger than 1600 characters. https://www.twilio.com/docs/api/errors/21617
+    # If the body is larger than 1600, Send the first 1600 chars.
+    original_body = body
+    body = body[0:1600]
+
+    # Create twilio message before calling twilio.
+    # If twilio message object created after message is sent to Twilio, race condition occurs.
+    # Some times Twilio calls status callback before the transaction is completed and message is saved.
+    twilio_message = TwilioMessage.objects.create(
```
Comment (efe) [inline django_sms_toolkit/tasks.py]: What will happen if the `TWILIO` comes to callback before filling the `message_sid`?

File: django_sms_toolkit/tasks.py (hunk 1 above)
Comment (alicertel) [reply django_sms_toolkit/tasks.py]: it's not responsibilty of the package.  You can turn this off in test mode.

File: django_sms_toolkit/tasks.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+@shared_task(retry_backoff=30, retry_kwargs={'max_retries': 3})
+def send_sms(from_number, to_number, body, recipient_id=None):
+    if not DJANGO_SMS_TOOLKIT_SETTINGS["SEND_SMS"]:
+        return 'Cannot send SMS because of environments settings.'
+
+    # Twilio does not accept body larger than 1600 characters. https://www.twilio.com/docs/api/errors/21617
+    # If the body is larger than 1600, Send the first 1600 chars.
+    original_body = body
+    body = body[0:1600]
```
[the remark this answers is not in the document]
Comment (alicertel) [reply django_sms_toolkit/tasks.py]: i didn't put this on purpose, if any of the project will need it in feature there should be PR for this.

File: django_sms_toolkit/tasks.py (hunk 2 above)
Comment (alicertel) [reply django_sms_toolkit/tasks.py]: `pk` of the object is identifier not the message_sid. The callback endpoint is unique per message.

Comment (efe) [reply django_sms_toolkit/tasks.py]: Oh, I see. Thank you.

File: django_sms_toolkit/models.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    def __str__(self):
+        return "Twilio message from {from_number} to {to_number}".format(from_number=self.from_number, to_number=self.to_number)
+
+    class Meta:
+        verbose_name = _('Twilio Message')
+        verbose_name_plural = _('Twilio Messages')
+        ordering = ["-creation_datetime"]
+
+    def save(self, *args, **kwargs):
+        self.clean()
```
Comment (cobang) [inline django_sms_toolkit/models.py]: I think we can use `full_clean` method.

File: django_sms_toolkit/views.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    def decorated_function(request, *args, **kwargs):
+        # Create an instance of the RequestValidator class
+        validator = RequestValidator(DJANGO_SMS_TOOLKIT_SETTINGS["TWILIO"]["AUTH_TOKEN"])
+
+        # Validate the request using its URL, POST data,
+        # and X-TWILIO-SIGNATURE header
+        request_valid = validator.validate(
+            request.build_absolute_uri(),
+            request.POST,
+            request.META.get('HTTP_X_TWILIO_SIGNATURE', ''))
```
Comment (cobang) [inline django_sms_toolkit/views.py]: I think, the last parenthesis should be on the next line.

File: django_sms_toolkit/models.py (hunk 4 above)
Comment (alicertel) [reply django_sms_toolkit/models.py]: Thank you fixed.

File: django_sms_toolkit/views.py (hunk 5 above)
Comment (alicertel) [reply django_sms_toolkit/views.py]: Fixed.

File: django_sms_toolkit/tasks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alicertel) [reply django_sms_toolkit/tasks.py]: Added `TRIM_LONG_BODY` settings just in case. Thank you for the feedback.

**Comment by alicertel** [comment]
I need to merge this. Please open an issue if you want something to be changed.

**Pull request merged**